descwl/analysis.py

```diff
@@ -521,55 +695,121 @@ def finalize(self,verbose,trace):
-            ('xmax',np.int32),
-            ('ymin',np.int32),
-            ('ymax',np.int32),
-            # Source properties.
-            ('f_disk', np.float32),
-            ('f_bulge', np.float32),
-            ('dx',np.float32),
-            ('dy',np.float32),
-            ('z',np.float32),
-            ('ab_mag',np.float32),
-            ('ri_color',np.float32),
-            ('flux',np.float32),
-            ('sigma_m',np.float32),
-            ('sigma_p',np.float32),
-            ('e1',np.float32),
-            ('e2',np.float32),
-            ('a',np.float32),
-            ('b',np.float32),
-            ('beta',np.float32),
-            ('psf_sigm',np.float32),
-            # Pixel-level properties.
-            ('purity',np.float32),
-            ('snr_sky',np.float32),
-            ('snr_iso',np.float32),
-            ('snr_grp',np.float32),
-            ('snr_isof',np.float32),
-            ('snr_grpf',np.float32),
-            ('ds',np.float32),
-            ('dg1',np.float32),
-            ('dg2',np.float32),
-            ('ds_grp',np.float32),
-            ('dg1_grp',np.float32),
-            ('dg2_grp',np.float32),
-            # HSM analysis results.
-            ('hsm_sigm',np.float32),
-            ('hsm_e1',np.float32),
-            ('hsm_e2',np.float32),
-            # Systematics fit results.
-            ('g1_fit',np.float32),
-            ('g2_fit',np.float32),
-            ])
+
+        if calculate_bias: 
+            data = np.empty(num_galaxies,dtype=[
+                ('db_id',np.int64),
+                ('grp_id',np.int64),
+                ('grp_size',np.int16),
+                ('grp_rank',np.int16),
+                ('visible',np.int16),
+                # Stamp bounding box.
+                ('xmin',np.int32),
+                ('xmax',np.int32),
+                ('ymin',np.int32),
+                ('ymax',np.int32),
+                # Source properties.
+                ('f_disk', np.float32),
+                ('f_bulge', np.float32),
+                ('dx',np.float32),
+                ('dy',np.float32),
+                ('z',np.float32),
+                ('ab_mag',np.float32),
+                ('ri_color',np.float32),
+                ('flux',np.float32),
+                ('sigma_m',np.float32),
+                ('sigma_p',np.float32),
+                ('e1',np.float32),
+                ('e2',np.float32),
+                ('a',np.float32),
+                ('b',np.float32),
+                ('beta',np.float32),
+                ('psf_sigm',np.float32),
+                # Pixel-level properties.
+                ('purity',np.float32),
+                ('snr_sky',np.float32),
+                ('snr_iso',np.float32),
+                ('snr_grp',np.float32),
+                ('snr_isof',np.float32),
+                ('snr_grpf',np.float32),
+                ('ds',np.float32),
+                ('dg1',np.float32),
+                ('dg2',np.float32),
+                ('bias_f', np.float32),
```

Comment:
There are a lot of duplicated dtype elements here.  You could avoid this by building a dtype array first, e.g.
```
dtype = [ ... ]
if calculate_bias:
    dtype.extend([ ... ])
data = np.empty(num_galaxies, dtype=dtype)
```